Global styles revisions: load unsaved revision item into the revisions preview

ramonjd opened the pull request:
## What?
Resolves: https://github.com/WordPress/gutenberg/issues/55866

### Before

https://github.com/WordPress/gutenberg/assets/6458278/bd2bb4f8-3771-4cfe-8d65-f9ee0e29178e

### After

https://github.com/WordPress/gutenberg/assets/6458278/3e6a1374-6047-4359-b66b-5d8c2917b0bd


## Why?

External changes to global styles in the editor, e.g., via the reset styles function of `useGlobalStylesReset()`, will create a new "unsaved" entry in the revisions list and replace the first item in the list. 

However the styles aren't loaded in the preview.

## How?
Always load the first style object into the revisions preview:

- the user hasn't selected any revision in the list, or
- the selected item is an unsaved revision, which the first one anyway, and
- the selected styles don't match the current editor styles
- added extra doc comments for context

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

### Testing Instructions for Keyboard

1. Navigate to Appearance -> Editor
2. Select a different style under Styles and Save (this creates a revision)
3. Open the revisions panel, check that the topmost, saved revision item is selected
4. Click on the Revisions symbol (a clock) and select Reset to defaults
5. A new unsaved changes revision item is created. Observe that the reset styles are loaded into the preview.
6. Select another revision and hit "Apply", now select the first item in the revisions list (the unsaved changes)
7. Click on the Revisions symbol (a clock) and select Reset to defaults (repeat step 4). Observe that the reset styles are loaded into the preview.

ramonjd committed:
Always load the first style object into the revisions preview:
- the user hasn't selected any revision in the list, or
- the selected item is an unsaved revision, which the first one anyway, and
- the selected styles don't match the current editor styles
- added extra doc comments for context

andrewserong commented on packages/edit-site/src/components/global-styles/screen-revisions/index.js:
```diff
@@ -55,13 +55,37 @@ function ScreenRevisions() {
+			This is required in case editor styles are changed outside the revisions panel,
+			e.g., via the reset styles function of useGlobalStylesReset().
+			See: https://github.com/WordPress/gutenberg/issues/55866
+		*/
```
Comment style nit: should this comment have an asterisk at the beginning of each line for consistency with the JS multi-line comments [guidelines](https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/#multi-line-comments)?

andrewserong commented on packages/edit-site/src/components/global-styles/screen-revisions/index.js:
```diff
@@ -55,13 +55,37 @@ function ScreenRevisions() {
+			setGlobalStylesRevision( revisions[ 0 ] );
 		}
-	}, [ editorCanvasContainerView ] );
+	}, [
+		editorCanvasContainerView,
+		shouldSelectFirstItem,
+		revisions[ 0 ],
+		setGlobalStylesRevision,
+	] );
```
This dependency array is complaining about some missing dependencies:

<img width="792" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/46a2a67c-12c9-4f39-a680-213a5cfdba6d">

To neaten things up, would it work to split out this new behaviour to a separate `useEffect`, that way the dependencies related to the `editorCanvasConatinerView` can be treated separately?

ramonjd replied:
Totally. I'll split them up to be neater.

andrewserong commented on packages/edit-site/src/components/global-styles/screen-revisions/index.js:
```diff
@@ -55,13 +55,37 @@ function ScreenRevisions() {
+			See: https://github.com/WordPress/gutenberg/issues/55866
+		*/
+		if ( shouldSelectFirstItem ) {
+			setGlobalStylesRevision( revisions[ 0 ] );
```
Does this also need to call `setSelectedRevisionId` as in `selectRevision` or is it enough to just update the revision?

ramonjd replied:
It's enough to update the revision as all we want to do is load the preview in the frame.

Having said that, I don't think it'd hurt to just call `selectRevision()` for consistency's sake.

👍🏻

andrewserong commented on packages/edit-site/src/components/global-styles/screen-revisions/index.js:
```diff
@@ -55,13 +55,37 @@ function ScreenRevisions() {
+	}, [
+		editorCanvasContainerView,
+		shouldSelectFirstItem,
+		revisions[ 0 ],
```
I think using `revisions` seems to cause the exhaustive-deps rule to expect `revisions` in the dependency array somewhere. If we only need the first element, I wonder if we can grab the first element outside of the useEffect via `const firstRevision = revisions[ 0 ]` and use that? Just a thought!

ramonjd replied:
This could work, thanks!

I'll try it.

andrewserong commented on packages/edit-site/src/components/global-styles/screen-revisions/index.js:
```diff
@@ -55,13 +55,37 @@ function ScreenRevisions() {
+		globalStylesRevision,
+		userConfig
+	);
+	const shouldSelectFirstItem =
+		! selectedRevisionMatchesEditorStyles &&
+		( ! selectedRevisionId || 'unsaved' === selectedRevisionId );
```
`areGlobalStyleConfigsEqual` checks that `globalStylesRevision` matches `userConfig`, however the `useEffect` updates only the `globalStylesRevision`. Could there ever be a case when the `useEffect` goes into an infinite loop if the updated `globalStylesRevision` doesn't match the `userConfig`?

Or to put it differently, in the `useEffect` do we need to also check that the current `globalStylesRevision` is not equal to the first item in the list of revisions, to avoid a potential infinite loop? Apologies if I'm overthinking this!

ramonjd replied:
> Or to put it differently, in the useEffect do we need to also check that the current globalStylesRevision is not equal to the first item in the list of revisions, to avoid a potential infinite loop? 

Not overthinking at all. `! selectedRevisionId || 'unsaved' === selectedRevisionId` implies that the first item is selected by default, but it wouldn't hurt to elaborate this defence. I'll have a play.

Thanks again for thinking about this one!

ramonjd replied:
> areGlobalStyleConfigsEqual checks that globalStylesRevision matches userConfig, however the useEffect updates only the globalStylesRevision. Could there ever be a case when the useEffect goes into an infinite loop if the updated globalStylesRevision doesn't match the userConfig?

Thinking about that made me realize that the `selectedRevisionId` is 100% redundant since it's stored in (and saved and the same time as) `globalStylesRevision`.

So, here's what I have:

```js
	const shouldSelectFirstItem =
                // If there is a revision to load in the first place,
                !! firstRevision?.id &&

                // AND if a revision is loaded in the preview pane that is
                // different from the current editor state (userConfig),
		! selectedRevisionMatchesEditorStyles &&

                // AND if there's no currently-selected revision (a user hasn't clicked on anything)
                // we can go ahead and select the first revision.
		 ! currentlySelectedRevision?.id;
```

What I'm trying to express here is that, if nothing has been selected, but for some reason the current state of the revisions pane doesn't reflect what's in the editor, let's fix that by selecting the first item in the list. 

So, if globalStylesRevision doesn't match the userConfig and there's no revision selected in the list the `useEffect` condition will pass and then select the first revision using the state setter `setCurrentlySelectedRevision`.

The next time it spins around it will fail because it's selected a revision and we've double checked that an id exists for `firstRevision?.id`. 

The only thing I can think of is if `firstRevision?.id` is falsey. But if that happens we've got bigger problems 😄

andrewserong replied:
Sounds good to me, thanks for laying that out!

andrewserong reviewed (commented):
This is testing nicely for me! Just left a few comments, mostly around readability of the `useEffect` and whether we can please the exhaustive deps linter. In practice it looks like the useEffect is firing when expected for me in manual testing 👍

Please let know if I'm overthinking any of my feedback here — I haven't looked at any code in the past couple of weeks so am a bit rusty in thinking through `useEffect`s 🙂

ramonjd committed:
Splitting useEffect
formatting comments

ramonjd committed:
selectedRevisionId is redundant since we're saving the same information in globalStylesRevision

ramonjd committed:
For clarity, rename:
`userConfig` to `currentEditorGlobalStyles`
`globalStylesRevision` to `currentlySelectedRevision`

Use `setCurrentlySelectedRevision` state setter in the useEffect callback as it's a stable ref

andrewserong commented on packages/edit-site/src/components/global-styles/screen-revisions/index.js:
```diff
@@ -77,17 +73,47 @@ function ScreenRevisions() {
+
+	useEffect( () => {
+		/*
+		 * Ensure that the first item is selected loaded into the preview pane
+		 * where no revision is selected and the selected styles don't match the current editor styles.
```
Totally optional tiny nits about wording:

```suggestion
		 * Ensure that the first item is selected and loaded into the preview pane
		 * when no revision is selected and the selected styles don't match the current editor styles.
```

andrewserong reviewed (approved):
Thanks for the thorough follow-ups @ramonjd and for thinking through all the changes. This is looking really good to me now, and I like how it's both simplified the state a bit, while including easier to read names that more closely match the behaviour we're examining 👍

Still testing nicely for me:

https://github.com/WordPress/gutenberg/assets/14988353/79aac76e-26be-453d-a4de-9bcd6e31bfd1

✅ Existing `useEffect` is still firing when expected to close out the panel and redirect to the root of global styles
✅ New `useEffect` fires only when expected and results in the unsaved revision item being updated as expected when resetting back to defaults

LGTM! ✨

ramonjd committed:
Update packages/edit-site/src/components/global-styles/screen-revisions/index.js

Co-authored-by: Andrew Serong <14988353+andrewserong@users.noreply.github.com>
